In [20]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules, fpmax, fpgrowth

In [21]:
# Load dataset
file_path = 'Online retail.xlsx'  # Replace with your dataset path
df = pd.read_excel(file_path)

In [22]:
df.describe()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
count                                                7500                                                                                                                                                                             
unique                                               5175                                                                                                                                                                             
top                                               cookies                                                                                                                                                                             
freq                                                  223

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 1 columns):
 #   Column                                                                                                                                                                                                                           Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                           --------------  ----- 
 0   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil  7500 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


In [24]:
df.columns =['Transactions']

In [25]:
print(df['Transactions'])

0                                  burgers,meatballs,eggs
1                                                 chutney
2                                          turkey,avocado
3       mineral water,milk,energy bar,whole wheat rice...
4                                          low fat yogurt
                              ...                        
7495                        butter,light mayo,fresh bread
7496    burgers,frozen vegetables,eggs,french fries,ma...
7497                                              chicken
7498                                   escalope,green tea
7499      eggs,frozen smoothie,yogurt cake,low fat yogurt
Name: Transactions, Length: 7500, dtype: object


In [26]:
# Assuming the dataset has a column 'Transactions' with items separated by commas
df['Transactions'] = df['Transactions'].fillna('').str.split(',')
transactions = df['Transactions'].tolist()

In [27]:
# Convert transactions into one-hot encoded DataFrame
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

In [29]:
df_encoded

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False    False              False      False    False        False   
1          False    False              False      False    False        False   
2          False    False              False      False     True        False   
3          False    False              False      False    False        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7495       False    False              False      False    False        False   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...    True   
3     False           False      False        False  ...   False   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7495  False           False      False        False  ...   False   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0              False        False       False             False   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7495           False        False       False             False   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False  False        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False              True  False        False     False  
4                 False             False  False        False     False  
...                 ...               ...    ...          ...       ...  
7495              False             False  False        False     False  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False         True     False  

[7500 rows x 120 columns]

In [35]:
# Apply Apriori algorithm
frequent_itemsets = apriori(df_encoded, min_support=0.01, use_colnames=True)

print(frequent_itemsets)

      support                                 itemsets
0    0.020267                                (almonds)
1    0.033200                                (avocado)
2    0.010800                         (barbecue sauce)
3    0.014267                              (black tea)
4    0.011467                             (body spray)
..        ...                                      ...
254  0.011067       (milk, mineral water, ground beef)
255  0.017067  (mineral water, spaghetti, ground beef)
256  0.015733         (milk, mineral water, spaghetti)
257  0.010267    (mineral water, olive oil, spaghetti)
258  0.011467     (pancakes, mineral water, spaghetti)

[259 rows x 2 columns]


In [48]:
# Generate association rules
# rules = association_rules(frequent_itemsets, metric="confidence ", min_threshold=1.0)
rules = association_rules(frequent_itemsets, metric='lift',min_threshold=0.7,num_itemsets=len(df_encoded))


In [53]:
# Display top rules
rules.sort_values('confidence',ascending=False)

antecedents                consequents  antecedent support  \
372        (eggs, ground beef)            (mineral water)            0.020000   
407        (milk, ground beef)            (mineral water)            0.022000   
347   (chocolate, ground beef)            (mineral water)            0.023067   
395  (milk, frozen vegetables)            (mineral water)            0.023600   
300                     (soup)            (mineral water)            0.050533   
..                         ...                        ...                 ...   
294            (mineral water)                 (red wine)            0.238267   
427            (mineral water)     (olive oil, spaghetti)            0.238267   
45             (mineral water)                  (cereals)            0.238267   
373            (mineral water)        (eggs, ground beef)            0.238267   
392            (mineral water)  (french fries, spaghetti)            0.238267   

     consequent support   support  confidence      lift  representativity  \
372            0.238267  0.010133    0.506667  2.126469               1.0   
407            0.238267  0.011067    0.503030  2.111207               1.0   
347            0.238267  0.010933    0.473988  1.989319               1.0   
395            0.238267  0.011067    0.468927  1.968075               1.0   
300            0.238267  0.023067    0.456464  1.915771               1.0   
..                  ...       ...         ...       ...               ...   
294            0.028133  0.010933    0.045887  1.631053               1.0   
427            0.022933  0.010267    0.043089  1.878880               1.0   
45             0.025733  0.010267    0.043089  1.674442               1.0   
373            0.020000  0.010133    0.042529  2.126469               1.0   
392            0.027600  0.010133    0.042529  1.540920               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
372  0.005368    1.544054       0.540548  0.040838   0.352354    0.274598  
407  0.005825    1.532756       0.538177  0.044409   0.347580    0.274738  
347  0.005437    1.448130       0.509058  0.043663   0.309454    0.259938  
395  0.005444    1.434328       0.503778  0.044125   0.302809    0.257687  
300  0.011026    1.401441       0.503458  0.086804   0.286449    0.276637  
..        ...         ...            ...       ...        ...         ...  
294  0.004230    1.018607       0.507919  0.042797   0.018268    0.217256  
427  0.004802    1.021063       0.614084  0.040914   0.020629    0.245382  
45   0.004135    1.018137       0.528776  0.040462   0.017814    0.221026  
373  0.005368    1.023530       0.695436  0.040838   0.022989    0.274598  
392  0.003557    1.015593       0.460840  0.039625   0.015353    0.204840  

[436 rows x 14 columns]

The association rules provide insights into items frequently purchased together. Each row represents a rule with key metrics such as support, confidence, lift, and additional advanced metrics.

Example Rules and Their Interpretations

Rule 1: (Eggs, Ground Beef) → (Mineral Water)

Confidence (50.67%):

If a customer buys eggs and ground beef, there's a 50.67% likelihood they'll also purchase mineral water.

Lift (2.13):

Customers who buy eggs and ground beef are 2.13 times more likely to buy mineral water compared to random customers.

Support (1.01%):

This combination occurs in 1.01% of transactions.

Insights into Customer Behavior

Mineral Water as a Key Item:

Mineral water frequently appears as a consequent, indicating it's commonly purchased alongside other items. 

Retailers could use this insight for:

Cross-promotion of mineral water with eggs, ground beef, or milk.

Strategic placement of mineral water near these products in stores.

Bundling Opportunities:

Products like "milk, frozen vegetables" and "ground beef" could be bundled together with mineral water to increase sales

Recommendations :

Promotions:

Offer discounts on mineral water when customers buy items like milk, eggs, or ground beef.
Create combo deals based on frequent antecedents.

Store Layout Optimization:

Place mineral water in sections with high traffic for eggs, milk, and ground beef.

Marketing Campaigns:

Highlight popular combinations in advertisements or loyalty programs.

**Interview Questions:**

1. What is Lift and Why is it Important in Association Rules?

Lift Definition:

Lift measures the strength of an association rule by comparing the observed frequency of itemsets to what would be expected if the items were independent.
 
Importance:
Lift > 1 indicates a positive correlation between items.
Lift < 1 indicates a negative correlation.
Lift = 1 implies no association (independence).

2. What are Support and Confidence? How Do You Calculate Them?

Support:

The proportion of transactions that contain a particular itemset

Confidence:

The probability that a transaction containing item A also contains item B.

3. What are Some Limitations or Challenges of Association Rule Mining?

High Dimensionality:

Large datasets with many items can lead to a combinatorial explosion of possible itemsets.

Parameter Sensitivity:

Results depend heavily on thresholds for support, confidence, and lift.

Interpretability:

Too many rules can be overwhelming, making it hard to extract actionable insights.

Rare Item Problem:

Rare items might not meet support thresholds but could have high lift and business value.

Data Quality Issues:

Missing or noisy data can affect rule accuracy.